### ISSUE TITLE
Bug with `replace_string` transformation

### ISSUE THREAD

**Issue by joshnck** [SigmaHQ/pySigma#262]
In our logs, we need to double escape because the fields `CommandLine` and `Image` are stored in Splunk such that `CommandLine=\\path\\to\\file.exe` and as such my search needs to be `CommandLine=\\\\path\\\\to\\\\file.exe`. I'm using the following transformation:

```yaml
  - id: one_to_four_slashes
    type: replace_string
    regex: "\\\\"
    replacement: "\\\\\\\\\\\\\\\\"
    rule_conditions:
      - type: logsource
        product: windows
    field_name_conditions:
      - type: include_fields
        fields:
          - Image
          - CommandLine
```
[Sigconverter](https://sigconverter.io/#backend=splunk&format=default&pipeline=&rule=dGl0bGU6IEdPT0QgUlVMRSAyCmlkOiA4MmE2NzE0Zi00ODk5LTRmMTYtOWMxZS05YTMzMzU0NGQ0YzMKc3RhdHVzOiB0ZXN0CmRlc2NyaXB0aW9uOiBmYWtlCmF1dGhvcjogZmFrZQpkYXRlOiAyMDIzLTA1LTE1CnJlZmVyZW5jZXM6CmxvZ3NvdXJjZToKICAgIGNhdGVnb3J5OiBwcm9jZXNzX2NyZWF0aW9uCiAgICBwcm9kdWN0OiB3aW5kb3dzCmRldGVjdGlvbjoKICAgIGNvbmRpdGlvbjogYWxsIG9mIHNlbGVjdGlvbl8qCiAgICBzZWxlY3Rpb25fY2xpOgogICAgICAgIENvbW1hbmRMaW5lfGNvbnRhaW5zOgogICAgICAgICAgICAtIFxBcHBEYXRhXFJvYW1pbmdcCiAgICAgICAgICAgIC0gXERlc2t0b3BcCiAgICAgICAgICAgIC0gXExvY2FsXFRlbXBcCiAgICAgICAgICAgIC0gXFBlcmZMb2dzXAogICAgICAgICAgICAtIFxVc2Vyc1xQdWJsaWNcCiAgICAgICAgICAgIC0gXFdpbmRvd3NcVGVtcFwKICAgICAgICAgICAgLSAkUmVjeWNsZS5CaW4KICAgIHNlbGVjdGlvbl9leHRlbnNpb246CiAgICAgICAgSW1hZ2V8Y29udGFpbnM6CiAgICAgICAgICAgIC0gXEFwcERhdGFcUm9hbWluZ1wKICAgICAgICAgICAgLSBcRGVza3RvcFwKICAgICAgICAgICAgLSBcTG9jYWxcVGVtcFwKICAgICAgICAgICAgLSBcUGVyZkxvZ3NcCiAgICAgICAgICAgIC0gXFVzZXJzXFB1YmxpY1wKICAgICAgICAgICAgLSBcV2luZG93c1xUZW1wXAogICAgICAgICAgICAtICRSZWN5Y2xlLkJpbgpmYWxzZXBvc2l0aXZlczogCiAgICAtIFVua25vd24KbGV2ZWw6IGxvdwp0YWdzOgogICAgLSBhdHRhY2suZGVmZW5zZS1ldmFzaW9uCiAgICAtIGF0dGFjay50MTAyNwo%3D&pipelineYml=cHJpb3JpdHk6IDIwCnRyYW5zZm9ybWF0aW9uczoKICAtIGlkOiBvbmVfdG9fZm91cl9zbGFzaGVzCiAgICB0eXBlOiByZXBsYWNlX3N0cmluZwogICAgcmVnZXg6ICJcXFxcIgogICAgcmVwbGFjZW1lbnQ6ICJcXFxcXFxcXFxcXFxcXFxcIgogICAgcnVsZV9jb25kaXRpb25zOgogICAgICAtIHR5cGU6IGxvZ3NvdXJjZQogICAgICAgIHByb2R1Y3Q6IHdpbmRvd3MKICAgIGZpZWxkX25hbWVfY29uZGl0aW9uczoKICAgICAgLSB0eXBlOiBpbmNsdWRlX2ZpZWxkcwogICAgICAgIGZpZWxkczoKICAgICAgICAgIC0gSW1hZ2UKICAgICAgICAgIC0gQ29tbWFuZExpbmUK)

Also in our logs full filepaths are not quite how they are in Sysmon - instead of `c:\\path\\to\\file.exe` they are `\\device\\harddiskvolume3\\path\\to\\file`. In some rules a full filepath is asserted or a `:\path\to\file` is asserted to mean a path starting at the root of a drive. I need two more transformations to get rid of this:

```yaml
  - id: replace_disk_name
    type: replace_string
    regex: "^[C-Z]:"
    replacement: "\\\\Device\\\\HarddiskVolume?\\\\"
    rule_conditions:
      - type: logsource
        product: windows
    field_name_conditions:
      - type: include_fields
        fields: Image
  - id: remove_colon_disk_name
    type: replace_string
    regex: ":"
    replacement: ""
    rule_conditions:
      - type: logsource
        product: windows
    field_name_conditions:
      - type: include_fields
        fields: Image
```
[sigconverter](https://sigconverter.io/#backend=splunk&format=default&pipeline=&rule=dGl0bGU6IEdPT0QgUlVMRSAyCmlkOiA4MmE2NzE0Zi00ODk5LTRmMTYtOWMxZS05YTMzMzU0NGQ0YzMKc3RhdHVzOiB0ZXN0CmRlc2NyaXB0aW9uOiBmYWtlCmF1dGhvcjogZmFrZQpkYXRlOiAyMDIzLTA1LTE1CnJlZmVyZW5jZXM6CmxvZ3NvdXJjZToKICAgIGNhdGVnb3J5OiBwcm9jZXNzX2NyZWF0aW9uCiAgICBwcm9kdWN0OiB3aW5kb3dzCmRldGVjdGlvbjoKICAgIGNvbmRpdGlvbjogYWxsIG9mIHNlbGVjdGlvbl8qCiAgICBzZWxlY3Rpb25fY2xpOgogICAgICAgIENvbW1hbmRMaW5lfGNvbnRhaW5zOgogICAgICAgICAgICAtIFxBcHBEYXRhXFJvYW1pbmdcCiAgICAgICAgICAgIC0gXERlc2t0b3BcCiAgICAgICAgICAgIC0gXExvY2FsXFRlbXBcCiAgICAgICAgICAgIC0gXFBlcmZMb2dzXAogICAgICAgICAgICAtIFxVc2Vyc1xQdWJsaWNcCiAgICAgICAgICAgIC0gXFdpbmRvd3NcVGVtcFwKICAgICAgICAgICAgLSAkUmVjeWNsZS5CaW4KICAgIHNlbGVjdGlvbl9leHRlbnNpb246CiAgICAgICAgSW1hZ2V8Y29udGFpbnM6CiAgICAgICAgICAgIC0gXEFwcERhdGFcUm9hbWluZ1wKICAgICAgICAgICAgLSBcRGVza3RvcFwKICAgICAgICAgICAgLSBcTG9jYWxcVGVtcFwKICAgICAgICAgICAgLSBcUGVyZkxvZ3NcCiAgICAgICAgICAgIC0gXFVzZXJzXFB1YmxpY1wKICAgICAgICAgICAgLSBcV2luZG93c1xUZW1wXAogICAgICAgICAgICAtICRSZWN5Y2xlLkJpbgpmYWxzZXBvc2l0aXZlczogCiAgICAtIFVua25vd24KbGV2ZWw6IGxvdwp0YWdzOgogICAgLSBhdHRhY2suZGVmZW5zZS1ldmFzaW9uCiAgICAtIGF0dGFjay50MTAyNwo%3D&pipelineYml=cHJpb3JpdHk6IDIwCnRyYW5zZm9ybWF0aW9uczoKICAtIGlkOiByZXBsYWNlX2Rpc2tfbmFtZQogICAgdHlwZTogcmVwbGFjZV9zdHJpbmcKICAgIHJlZ2V4OiAiXltDLVpdOiIKICAgIHJlcGxhY2VtZW50OiAiXFxcXERldmljZVxcXFxIYXJkZGlza1ZvbHVtZT9cXFxcIgogICAgcnVsZV9jb25kaXRpb25zOgogICAgICAtIHR5cGU6IGxvZ3NvdXJjZQogICAgICAgIHByb2R1Y3Q6IHdpbmRvd3MKICAgIGZpZWxkX25hbWVfY29uZGl0aW9uczoKICAgICAgLSB0eXBlOiBpbmNsdWRlX2ZpZWxkcwogICAgICAgIGZpZWxkczogSW1hZ2UKICAtIGlkOiByZW1vdmVfY29sb25fZGlza19uYW1lCiAgICB0eXBlOiByZXBsYWNlX3N0cmluZwogICAgcmVnZXg6ICI6IgogICAgcmVwbGFjZW1lbnQ6ICIiCiAgICBydWxlX2NvbmRpdGlvbnM6CiAgICAgIC0gdHlwZTogbG9nc291cmNlCiAgICAgICAgcHJvZHVjdDogd2luZG93cwogICAgZmllbGRfbmFtZV9jb25kaXRpb25zOgogICAgICAtIHR5cGU6IGluY2x1ZGVfZmllbGRzCiAgICAgICAgZmllbGRzOiBJbWFnZQo%3D)

Now you'll already see some weird things happening here - but before I continue let me combine all 3:
```yaml
priority: 20
transformations:
  - id: one_to_four_slashes
    type: replace_string
    regex: "\\\\"
    replacement: "\\\\\\\\\\\\\\\\"
    rule_conditions:
      - type: logsource
        product: windows
    field_name_conditions:
      - type: include_fields
        fields:
          - Image
          - CommandLine
  - id: replace_disk_name
    type: replace_string
    regex: "^[C-Z]:"
    replacement: "\\\\Device\\\\HarddiskVolume?\\\\"
    rule_conditions:
      - type: logsource
        product: windows
    field_name_conditions:
      - type: include_fields
        fields: Image
  - id: remove_colon_disk_name
    type: replace_string
    regex: ":"
    replacement: ""
    rule_conditions:
      - type: logsource
        product: windows
    field_name_conditions:
      - type: include_fields
        fields: Image
```

[sigconverter](https://sigconverter.io/#backend=splunk&format=default&pipeline=&rule=dGl0bGU6IEdPT0QgUlVMRSAyCmlkOiA4MmE2NzE0Zi00ODk5LTRmMTYtOWMxZS05YTMzMzU0NGQ0YzMKc3RhdHVzOiB0ZXN0CmRlc2NyaXB0aW9uOiBmYWtlCmF1dGhvcjogZmFrZQpkYXRlOiAyMDIzLTA1LTE1CnJlZmVyZW5jZXM6CmxvZ3NvdXJjZToKICAgIGNhdGVnb3J5OiBwcm9jZXNzX2NyZWF0aW9uCiAgICBwcm9kdWN0OiB3aW5kb3dzCmRldGVjdGlvbjoKICAgIGNvbmRpdGlvbjogYWxsIG9mIHNlbGVjdGlvbl8qCiAgICBzZWxlY3Rpb25fY2xpOgogICAgICAgIENvbW1hbmRMaW5lfGNvbnRhaW5zOgogICAgICAgICAgICAtIFxBcHBEYXRhXFJvYW1pbmdcCiAgICAgICAgICAgIC0gXERlc2t0b3BcCiAgICAgICAgICAgIC0gXExvY2FsXFRlbXBcCiAgICAgICAgICAgIC0gXFBlcmZMb2dzXAogICAgICAgICAgICAtIFxVc2Vyc1xQdWJsaWNcCiAgICAgICAgICAgIC0gXFdpbmRvd3NcVGVtcFwKICAgICAgICAgICAgLSAkUmVjeWNsZS5CaW4KICAgIHNlbGVjdGlvbl9leHRlbnNpb246CiAgICAgICAgSW1hZ2V8Y29udGFpbnM6CiAgICAgICAgICAgIC0gXEFwcERhdGFcUm9hbWluZ1wKICAgICAgICAgICAgLSBcRGVza3RvcFwKICAgICAgICAgICAgLSBcTG9jYWxcVGVtcFwKICAgICAgICAgICAgLSBcUGVyZkxvZ3NcCiAgICAgICAgICAgIC0gXFVzZXJzXFB1YmxpY1wKICAgICAgICAgICAgLSBcV2luZG93c1xUZW1wXAogICAgICAgICAgICAtICRSZWN5Y2xlLkJpbgpmYWxzZXBvc2l0aXZlczogCiAgICAtIFVua25vd24KbGV2ZWw6IGxvdwp0YWdzOgogICAgLSBhdHRhY2suZGVmZW5zZS1ldmFzaW9uCiAgICAtIGF0dGFjay50MTAyNwo%3D&pipelineYml=cHJpb3JpdHk6IDIwCnRyYW5zZm9ybWF0aW9uczoKICAtIGlkOiBvbmVfdG9fZm91cl9zbGFzaGVzCiAgICB0eXBlOiByZXBsYWNlX3N0cmluZwogICAgcmVnZXg6ICJcXFxcIgogICAgcmVwbGFjZW1lbnQ6ICJcXFxcXFxcXFxcXFxcXFxcIgogICAgcnVsZV9jb25kaXRpb25zOgogICAgICAtIHR5cGU6IGxvZ3NvdXJjZQogICAgICAgIHByb2R1Y3Q6IHdpbmRvd3MKICAgIGZpZWxkX25hbWVfY29uZGl0aW9uczoKICAgICAgLSB0eXBlOiBpbmNsdWRlX2ZpZWxkcwogICAgICAgIGZpZWxkczoKICAgICAgICAgIC0gSW1hZ2UKICAgICAgICAgIC0gQ29tbWFuZExpbmUKICAtIGlkOiByZXBsYWNlX2Rpc2tfbmFtZQogICAgdHlwZTogcmVwbGFjZV9zdHJpbmcKICAgIHJlZ2V4OiAiXltDLVpdOiIKICAgIHJlcGxhY2VtZW50OiAiXFxcXERldmljZVxcXFxIYXJkZGlza1ZvbHVtZT9cXFxcIgogICAgcnVsZV9jb25kaXRpb25zOgogICAgICAtIHR5cGU6IGxvZ3NvdXJjZQogICAgICAgIHByb2R1Y3Q6IHdpbmRvd3MKICAgIGZpZWxkX25hbWVfY29uZGl0aW9uczoKICAgICAgLSB0eXBlOiBpbmNsdWRlX2ZpZWxkcwogICAgICAgIGZpZWxkczogSW1hZ2UKICAtIGlkOiByZW1vdmVfY29sb25fZGlza19uYW1lCiAgICB0eXBlOiByZXBsYWNlX3N0cmluZwogICAgcmVnZXg6ICI6IgogICAgcmVwbGFjZW1lbnQ6ICIiCiAgICBydWxlX2NvbmRpdGlvbnM6CiAgICAgIC0gdHlwZTogbG9nc291cmNlCiAgICAgICAgcHJvZHVjdDogd2luZG93cwogICAgZmllbGRfbmFtZV9jb25kaXRpb25zOgogICAgICAtIHR5cGU6IGluY2x1ZGVfZmllbGRzCiAgICAgICAgZmllbGRzOiBJbWFnZQo%3D)

Each time I add a new `replace_string` modifying `Image` it removes another slash at the end! Even though the regex for the 2nd and 3rd transformation do not match!

Transformation with all 3 `replace_string`
```
CommandLine IN ("*\\\\AppData\\\\Roaming\\\\*", "*\\\\Desktop\\\\*", "*\\\\Local\\\\Temp\\\\*", "*\\\\PerfLogs\\\\*", "*\\\\Users\\\\Public\\\\*", "*\\\\Windows\\\\Temp\\\\*", "*$Recycle.Bin*") Image IN ("*\\AppData\\Roaming\*", "*\\Desktop\*", "*\\Local\\Temp\*", "*\\PerfLogs\*", "*\\Users\\Public\*", "*\\Windows\\Temp\*", "*$Recycle.Bin*")
```
Transformation with `remove_colon_disk_name` removed:
```
CommandLine IN ("*\\\\AppData\\\\Roaming\\\\*", "*\\\\Desktop\\\\*", "*\\\\Local\\\\Temp\\\\*", "*\\\\PerfLogs\\\\*", "*\\\\Users\\\\Public\\\\*", "*\\\\Windows\\\\Temp\\\\*", "*$Recycle.Bin*") Image IN ("*\\AppData\\Roaming\\*", "*\\Desktop\\*", "*\\Local\\Temp\\*", "*\\PerfLogs\\*", "*\\Users\\Public\\*", "*\\Windows\\Temp\\*", "*$Recycle.Bin*")
```
Transformation with only the one-to-four slashes replace string (the one that I need to work properly above all else)
```
CommandLine IN ("*\\\\AppData\\\\Roaming\\\\*", "*\\\\Desktop\\\\*", "*\\\\Local\\\\Temp\\\\*", "*\\\\PerfLogs\\\\*", "*\\\\Users\\\\Public\\\\*", "*\\\\Windows\\\\Temp\\\\*", "*$Recycle.Bin*") Image IN ("*\\\\AppData\\\\Roaming\\\\*", "*\\\\Desktop\\\\*", "*\\\\Local\\\\Temp\\\\*", "*\\\\PerfLogs\\\\*", "*\\\\Users\\\\Public\\\\*", "*\\\\Windows\\\\Temp\\\\*", "*$Recycle.Bin*")
```

**Comment by joshnck** [comment]
I haven't had a chance to fully trace this in the code, but a hint is that it only happens with the `\` character. Using `xxxxxx` as the initial replacement does not yield the same results: [sigconverter.io](https://sigconverter.io/#backend=splunk&format=default&pipeline=&rule=dGl0bGU6IEdPT0QgUlVMRSAyCmlkOiA4MmE2NzE0Zi00ODk5LTRmMTYtOWMxZS05YTMzMzU0NGQ0YzMKc3RhdHVzOiB0ZXN0CmRlc2NyaXB0aW9uOiBmYWtlCmF1dGhvcjogZmFrZQpkYXRlOiAyMDIzLTA1LTE1CnJlZmVyZW5jZXM6CmxvZ3NvdXJjZToKICAgIGNhdGVnb3J5OiBwcm9jZXNzX2NyZWF0aW9uCiAgICBwcm9kdWN0OiB3aW5kb3dzCmRldGVjdGlvbjoKICAgIGNvbmRpdGlvbjogYWxsIG9mIHNlbGVjdGlvbl8qCiAgICBzZWxlY3Rpb25fY2xpOgogICAgICAgIENvbW1hbmRMaW5lfGNvbnRhaW5zOgogICAgICAgICAgICAtIFxBcHBEYXRhXFJvYW1pbmdcCiAgICAgICAgICAgIC0gXERlc2t0b3BcCiAgICAgICAgICAgIC0gXExvY2FsXFRlbXBcCiAgICAgICAgICAgIC0gXFBlcmZMb2dzXAogICAgICAgICAgICAtIFxVc2Vyc1xQdWJsaWNcCiAgICAgICAgICAgIC0gXFdpbmRvd3NcVGVtcFwKICAgICAgICAgICAgLSAkUmVjeWNsZS5CaW4KICAgIHNlbGVjdGlvbl9leHRlbnNpb246CiAgICAgICAgSW1hZ2V8Y29udGFpbnM6CiAgICAgICAgICAgIC0gXEFwcERhdGFcUm9hbWluZ1wKICAgICAgICAgICAgLSBcRGVza3RvcFwKICAgICAgICAgICAgLSBcTG9jYWxcVGVtcFwKICAgICAgICAgICAgLSBcUGVyZkxvZ3NcCiAgICAgICAgICAgIC0gXFVzZXJzXFB1YmxpY1wKICAgICAgICAgICAgLSBcV2luZG93c1xUZW1wXAogICAgICAgICAgICAtICRSZWN5Y2xlLkJpbgpmYWxzZXBvc2l0aXZlczogCiAgICAtIFVua25vd24KbGV2ZWw6IGxvdwp0YWdzOgogICAgLSBhdHRhY2suZGVmZW5zZS1ldmFzaW9uCiAgICAtIGF0dGFjay50MTAyNwo%3D&pipelineYml=cHJpb3JpdHk6IDIwCnRyYW5zZm9ybWF0aW9uczoKICAtIGlkOiBvbmVfdG9fZm91cl9zbGFzaGVzCiAgICB0eXBlOiByZXBsYWNlX3N0cmluZwogICAgcmVnZXg6ICJcXFxcIgogICAgcmVwbGFjZW1lbnQ6ICJ4eHh4eHh4eHh4eHgiCiAgICBydWxlX2NvbmRpdGlvbnM6CiAgICAgIC0gdHlwZTogbG9nc291cmNlCiAgICAgICAgcHJvZHVjdDogd2luZG93cwogICAgZmllbGRfbmFtZV9jb25kaXRpb25zOgogICAgICAtIHR5cGU6IGluY2x1ZGVfZmllbGRzCiAgICAgICAgZmllbGRzOgogICAgICAgICAgLSBJbWFnZQogICAgICAgICAgLSBDb21tYW5kTGluZQogIC0gaWQ6IHJlcGxhY2VfZGlza19uYW1lCiAgICB0eXBlOiByZXBsYWNlX3N0cmluZwogICAgcmVnZXg6ICJeW0MtWl06IgogICAgcmVwbGFjZW1lbnQ6ICJcXFxcRGV2aWNlXFxcXEhhcmRkaXNrVm9sdW1lP1xcXFwiCiAgICBydWxlX2NvbmRpdGlvbnM6CiAgICAgIC0gdHlwZTogbG9nc291cmNlCiAgICAgICAgcHJvZHVjdDogd2luZG93cwogICAgZmllbGRfbmFtZV9jb25kaXRpb25zOgogICAgICAtIHR5cGU6IGluY2x1ZGVfZmllbGRzCiAgICAgICAgZmllbGRzOiBJbWFnZQogIC0gaWQ6IHJlbW92ZV9jb2xvbl9kaXNrX25hbWUKICAgIHR5cGU6IHJlcGxhY2Vfc3RyaW5nCiAgICByZWdleDogIjoiCiAgICByZXBsYWNlbWVudDogIiIKICAgIHJ1bGVfY29uZGl0aW9uczoKICAgICAgLSB0eXBlOiBsb2dzb3VyY2UKICAgICAgICBwcm9kdWN0OiB3aW5kb3dzCiAgICBmaWVsZF9uYW1lX2NvbmRpdGlvbnM6CiAgICAgIC0gdHlwZTogaW5jbHVkZV9maWVsZHMKICAgICAgICBmaWVsZHM6IEltYWdlCg%3D%3D)

**Comment by thomaspatzke** [comment]
Will look deeper into it on the weekend but I already have an idea what the problem could be. Escape and wildcard characters get some special treatment because they are out of band and not contained in the plating string.

**Comment by martinspielmann** [comment]
I wonder how much money and effort has been wasted over the years because of the brilliant idea of making both escape character and file separator the same thing on Windows😭 Anyone out there to sue them, Delta Air maybe?

**Comment by thomaspatzke** [comment]
Fixed in v0.11.11.
